File: app/src/main/java/at/privatepilotserver/server/http.js
const http = require('http');
const { getPublicKey } = require('./encryption');
const { registerUser } = require('./userAuth');

let registerUserEndpointEnabled = true; // server.js changes value

function startLoginServer(port) {
    const server = http.createServer((req, res) => {
        if (req.url === '/public-key' && req.method === 'GET') {
            res.writeHead(200, { 'Content-Type': 'text/plain' });
            res.end(getPublicKey());
        } else if (registerUserEndpointEnabled && req.method === 'POST' && req.url === '/register-user') {
            let name = req.headers['name']
            let data = '';
            req.on('data', (chunk) => {
                data += chunk;
            });
            req.on('end', () => {
                try {
                    registerUser(name, data);

                    res.writeHead(200, { 'Content-Type': 'text/plain' });
                    res.end('User registered successfully');
                } catch (error) {
                    console.error('Error reading request body:', error);
                    res.writeHead(400, { 'Content-Type': 'text/plain' });
                    res.end('Bad Request');
                }
            });
        } else {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('Not Found');
        }
    });

    server.listen(port, () => {
        console.log(`Public Key Server is running on port ${port}`);
    });

    return server; // Return the server instance
}

module.exports = {
    startLoginServer,
    enableRegisterUserEndpoint: () => (registerUserEndpointEnabled = true),
    disableRegisterUserEndpoint: () => (registerUserEndpointEnabled = false)
};

File: app/src/main/java/at/privatepilot/server/Http.kt
package at.privatepilot.server

import android.content.Context
import io.ktor.application.call
import io.ktor.application.install
import io.ktor.features.ContentNegotiation
import io.ktor.gson.gson
import io.ktor.http.HttpStatusCode
import io.ktor.request.receiveText
import io.ktor.response.respond
import io.ktor.response.respondText
import io.ktor.routing.get
import io.ktor.routing.post
import io.ktor.routing.routing
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty
import io.ktor.server.netty.NettyApplicationEngine
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class Http(private val keyPort: Int, private val registerPort: Int, val encryption: Encryption, val context: Context) {

    private var registerServer: NettyApplicationEngine? = null
    private val userAuth = UserAuth(context)

    init {
        // Start the Ktor server for /public-key in a coroutine
        GlobalScope.launch(Dispatchers.IO) {
            embeddedServer(Netty, keyPort) {
                install(ContentNegotiation) {
                    gson {}
                }
                routing {
                    get("/public-key") {
                        call.respondText(encryption.getPublicKey())
                    }
                    get("/ip") {
                        call.respondText(NetworkManager.getInternetIpAddress())
                    }
                }
            }.start(wait = true)
        }
    }

    fun startRegisterUserServer(){
        // Start the Ktor server for /register-user in a coroutine
        GlobalScope.launch(Dispatchers.IO) {
            registerServer = embeddedServer(Netty, registerPort) {
                install(ContentNegotiation) {
                    gson {}
                }
                routing {
                    get("/") {
                        call.respond("Attention register Server is Open!")
                    }
                    post("/register-user") {
                        val name = call.request.headers["name"] ?: ""
                        val ip = call.request.headers["ip"] ?: ""
                        val data = call.receiveText()

                        try {
                            userAuth.registerUser(name, ip, data)
                            call.respondText("User registered successfully")
                        } catch (e: Exception) {
                            call.respond(HttpStatusCode.BadRequest, "Bad Request")
                        }
                    }
                }
            }
            registerServer?.start(wait = true)
        }
    }

    fun stopRegisterUserServer() {
        registerServer?.stop(1000, 5000)
    }
}
